imap_processing/ultra/l0/decom_ultra.py

```diff
@@ -0,0 +1,192 @@
+
+    Returns
+    -------
+    list
+        A list of decompressed values.
+
+    Note: Equations from Section 1.2.1.1 Data Compression and Decompression Algorithms
+    in Ultra_algorithm_doc_rev2.pdf.
+    """
+    current_position = 0
+    decompressed_values = []
+
+    while current_position < len(binary):
+        # Read the width of the block
+        width, current_position = read_n_bits(binary, width_bit, current_position)
+        # If width is 0 or None, we don't have enough bits left
+        if width is None or len(decompressed_values) >= \
+                UltraParams.ULTRA_IMG_RATES.value.len_array:
+            print('hi')
+            break
+
+        # For each block, read 16 values of the given width
+        for _ in range(block):
+            # Ensure there are enough bits left to read the width
+            if len(binary) - current_position < width:
+                break
+
+            value, current_position = read_n_bits(binary, width, current_position)
+
+            # Log decompression
+            decompressed_values.append(log_decompression(value))
+
+    return decompressed_values
+
+
+def decom_ultra_packets(packet_file: str, xtce: str):
+    """
+    Unpack and decode ultra packets using CCSDS format and XTCE packet definitions.
+
+    Parameters
+    ----------
+    packet_file : str
+        Path to the CCSDS data packet file.
+    xtce : str
+        Path to the XTCE packet definition file.
+
+    Returns
+    -------
+    xr.Dataset
+        A dataset containing the decoded data fields with 'time' as the coordinating
+        dimension.
+    """
+
+    packets = decom.decom_packets(packet_file, xtce)
+
+    met_data, science_id, spin_data, abortflag_data, startdelay_data, fastdata_00 = \
+        [], [], [], [], [], []
+
+    for packet in packets:
+        if packet.header['PKT_APID'].derived_value == \
+                UltraParams.ULTRA_IMG_RATES.value.apid:
+            met_data.append(packet.data['SHCOARSE'].derived_value)
+            science_id.append(packet.data['SID'].derived_value)
+            spin_data.append(packet.data['SPIN'].derived_value)
+            abortflag_data.append(packet.data['ABORTFLAG'].derived_value)
+            startdelay_data.append(packet.data['STARTDELAY'].derived_value)
+            decompressed_data = decompress_binary(packet.data['FASTDATA_00'].raw_value,
+                                                  UltraParams.ULTRA_IMG_RATES.value.width,
+                                                  UltraParams.ULTRA_IMG_RATES.value.block)
+            fastdata_00.append(decompressed_data)
+
+    array_data = np.array(fastdata_00)
+
+    ds = xr.Dataset({
+        'science_id': ('epoch', science_id),
+        'spin_data': ('epoch', spin_data),
+        'abortflag_data': ('epoch', abortflag_data),
+        'startdelay_data': ('epoch', startdelay_data),
+        'fastdata_00': (('epoch', 'index'), array_data)
+    }, coords={
+        'epoch': met_data,
```

Comment:
@bryan-harter Does it matter is it's lower case or not? Asking for my future self.